modules/RoutingContext.js

```diff
@@ -5,15 +6,28 @@ import getRouteParams from './getRouteParams'
 
 const { array, func, object } = React.PropTypes
 
-/**
- * A <RoutingContext> renders the component tree for a given router state
- * and sets the history object and the current location in context.
- */
 class RoutingContext extends Component {
 
   getChildContext() {
     const { history, location } = this.props
-    return { history, location }
+    const router = {
+      push(...args) {
+        history.push(...args)
+      },
+      replace(...args) {
+        history.replace(...args)
+      },
+      addRouteLeaveHook(...args) {
+        return history.listenBeforeLeavingRoute(...args)
+      },
+      isActive(...args) {
+        return history.isActive(...args)
+      }
+    }
+    return deprecateObjectProperties({ history, location, router }, {
```

Comment:
For `warning` and `invariant`, it's a Babel plugin that strips out the warning messages in production builds.

But I feel like any effort we might want to put into that sort of thing would be better spent on providing codemods for our users to shift to the new API, rather than deprecation ergonomics on our end.